ratelimit.go

```diff
@@ -48,28 +45,15 @@ type Clock interface {
 	Sleep(time.Duration)
 }
 
-type state struct {
-	last     time.Time
-	sleepFor time.Duration
-}
-
-type limiter struct {
-	state unsafe.Pointer
-	//lint:ignore U1000 Padding is unused but it is crucial to maintain performance
-	// of this rate limiter in case of collocation with other frequently accessed memory.
-	padding [56]byte // cache line size - state pointer size = 64 - 8; created to avoid false sharing.
-
-	perRequest time.Duration
-	maxSlack   time.Duration
-	clock      Clock
-}
-
 // config configures a limiter.
 type config struct {
 	maxSlack time.Duration
 	clock    Clock
 }
 
+// New returns a Limiter that will limit to the given RPS.
+var New = newAtomicBased
```

Comment:
instead of using a global which can be avoided, can we use a wrapper function instead, and export the separate constructors? this is so that we can return pointers to structs when possible instead of returning interfaces. i.e.
```
func New(...) Limiter  {
  return NewAtomicLimiter(...)
}

func NewAtomicLimiter(...) *AtomicLimiter {...}

func NewMutexLimiter(...) *MutexLimiter {...}
```

Reply:
Makes sense - done.

There's no immediate benefit of returning (pointers to) structs here though, right? It could make it easier to test some internals of the structs, but that's it - or am I missing something?

Reply:
Testing is also part of the reason, but it's just good practice in general - by returning an interface instead of a struct, you limit the interface that the exported object is able to implement to just the one you've returned, and it becomes really hard to e.g. add a `.Write()` or `.String()` to the object without modifying the interface.